### DISCUSSION TITLE
Fairlearn Roadmap Ideas

### DISCUSSION THREAD

**Discussion by hildeweerts** [fairlearn/fairlearn#696] [Ideas]
# Roadmap Ideas

#### 2.1 Plan
* Need to create a list of topics we would want to teach in educational materials
* Need to create a scaffold or organizational structure/taxonomy etc. of topics, possibly a sequenced curriculum or textbook
* Strategy or path for creating learning objectives and learning materials for Fairlearn

#### 2.2 Use Cases and Example Notebooks
* Create use cases and/or example notebooks that demonstrate and/or teach fairness concepts in sociotechnical contexts
* Including testimonials and/or scenarios of groups who have already used Fairlearn in their work - more detailed than a 1-paragraph summary, more accessible (e.g., notebook) than a white paper

#### 2.3 General educational materials and resources
* Topics
   * Abstraction traps (see [abstraction traps](https://dl.acm.org/doi/10.1145/3287560.3287598))
   * Construct validity and reliability (see [measurement and fairness](https://arxiv.org/abs/1912.05511))
   * Long-term fairness
   * ...
* Toolkits
   * Toolkit to create datasheets, as in Google’s model cards toolkit
   * Toolkit to use fairness metric in model cards
   * Toolkit to help data scientists think about construct validity and reliability
   * ...
  
### 2. Documentation
* Improve API docs/User guide. Ensure documentation is up-to-date, explains things well, connects to external resources that give context

### 3. New Technical Features
* Integrate some of the features that are in use in other fairness toolkits (e.g., IBM AI Fairness 360, scikit-fairness), e.g., re-labeling, re-weighting (see https://github.com/fairlearn/fairlearn/discussions/680)
* Add error bars to MetricFrame
* Add counterfactual evaluations (i.e., how to measure fairness with censored data in problems such as recidivism, admissions, [child welfare screening](https://arxiv.org/pdf/1909.00066.pdf))
* Include metrics for counterfactual fairness, including causal modeling

### 4. Website Revision
* Revise website structure (See e.g., https://github.com/fairlearn/fairlearn/discussions/682 and https://github.com/fairlearn/fairlearn/discussions/685)
* 1-paragraph testimonials for funding

**Comment by trivenigandhi** [comment]
This roadmap looks awesome! I am obviously a big fan of the new technical features around pre-processing techniques :) 
In terms of the education materials I would suggest we provide a lot of good material on how to understand the socio-technical relevance of Responsible AI, how the entire process of developing AI is at risk and _then_ how it can be addressed from the ML fairness perspective. I think the potential value add of educational materials should be about making the socio-techinical considerations more real and relevant for highly technical profiles - let's build a bridge to the non-tech considerations.

**Reply by hildeweerts** [reply]
Are there any specific pre-processing techniques you'd like to see in the library?

I 100% agree with trying to build bridges. I wonder what's the best way to "organize" the educational materials. I personally like the idea of mapping different considerations/concepts/toolkits to different stages of the ML development process, because that's a structure practitioners are already familiar with. But I'm curious to hear what other folks think!

**Reply by hildeweerts** [reply]
Something I think would be super valuable for exemplifying some of the fairness issues and trade-offs, is to create interactive visuals like this one: https://www.technologyreview.com/2019/10/17/75285/ai-fairer-than-judge-criminal-risk-assessment-algorithm/
